File: Contest 3 Task 2A (all files)/ArrayGenerator.cs
#include <vector>
#include <algorithm>
#include <random>
#include <chrono>

class ArrayGenerator
{
    private:
    std::vector<int> baseArray;
    std::mt19937 rng; // Генератор случайных чисел

    public:
    ArrayGenerator() : rng(std::chrono::steady_clock::now().time_since_epoch().count())
    {
        // Создание базового массива длины 10000 со случайными числами
        baseArray.resize(10000);
        std::uniform_int_distribution<int> dist(0, 6000);
        for (auto & elem : baseArray)
        {
            elem = dist(rng);
        }
    }

    // Генерация массива случайных значений заданного размера
    std::vector<int> getRandomArray(size_t size)
    {
        return std::vector<int>(baseArray.begin(), baseArray.begin() + size);
    }

    // Генерация массива, отсортированного в обратном порядке
    std::vector<int> getReverseSortedArray(size_t size)
    {
        std::vector<int> arr = getRandomArray(size);
        std::sort(arr.rbegin(), arr.rend());
        return arr;
    }

    // Генерация "почти отсортированного" массива
    std::vector<int> getNearlySortedArray(size_t size, int swaps = 10)
    {
        std::vector<int> arr = getRandomArray(size);
        std::sort(arr.begin(), arr.end());
        for (int i = 0; i < swaps; ++i)
        {
            size_t idx1 = rng() % size;
            size_t idx2 = rng() % size;
            std::swap(arr[idx1], arr[idx2]);
        }
        return arr;
    }
};

File: Contest 3 Task 2A (all files)/ArrayGenerator.cpp
#include "ArrayGenerator.h"
#include <algorithm>
#include <random>
#include <ctime>

void ArrayGenerator::generateBaseArray() {
  std::mt19937 generator(static_cast<unsigned>(std::time(nullptr)));
  std::uniform_int_distribution<int> distribution(minValue, maxValue);

  for (int i = 0; i < maxArraySize; ++i) {
    baseArray.push_back(distribution(generator));
  }
}

ArrayGenerator::ArrayGenerator() : maxArraySize(10000), minValue(0), maxValue(6000) {
  generateBaseArray();
}

std::vector<int> ArrayGenerator::getRandomArray(int size) {
  if (size > maxArraySize) {
    size = maxArraySize;
  }
  return std::vector<int>(baseArray.begin(), baseArray.begin() + size);
}

std::vector<int> ArrayGenerator::getReversedArray(int size) {
  if (size > maxArraySize) {
    size = maxArraySize;
  }
  std::vector<int> reversedArray(baseArray.begin(), baseArray.begin() + size);
  std::sort(reversedArray.rbegin(), reversedArray.rend());
  return reversedArray;
}

std::vector<int> ArrayGenerator::getNearlySortedArray(int size, int swaps) {
  if (size > maxArraySize) {
    size = maxArraySize;
  }
  std::vector<int> nearlySortedArray(baseArray.begin(), baseArray.begin() + size);
  std::sort(nearlySortedArray.begin(), nearlySortedArray.end());

  std::mt19937 generator(static_cast<unsigned>(std::time(nullptr)));
  std::uniform_int_distribution<int> distribution(0, size - 1);

  for (int i = 0; i < swaps; ++i) {
    std::swap(nearlySortedArray[distribution(generator)], nearlySortedArray[distribution(generator)]);
  }

  return nearlySortedArray;
}
